Java: make null-pointer and similar checks fatal

smowton opened the pull request:
These previously used code_assertt, which asserts but then continues -- this
changes them to use assert(condition); assume(condition), which ensures subsequent
code is not executed in conditions which are in practice impossible.

With --java-throw-runtime-exceptions this is irrelevant, as a real exception is raised
and the conditional GOTO guarding the throw has the same effect as the assume.

smowton committed:
Java: make null-pointer and similar checks fatal

These previously used code_assertt, which asserts but then continues -- this
changes them to use assert(condition); assume(condition), which ensures subsequent
code is not executed in conditions which are in practice impossible.

With --java-throw-runtime-exceptions this is irrelevant, as a real exception is raised
and the conditional GOTO guarding the throw has the same effect as the assume.

smowton commented:
This will need a test-gen bump once test-gen#1794 is in.

tautschnig commented on src/java_bytecode/java_bytecode_instrument.cpp:
```diff
@@ -82,6 +83,21 @@ const std::vector<std::string> exception_needed_classes = {
+/// \param loc: original source location
+/// \param comment: human-readable assertion description
+/// \param property_class: assertion property class
+static source_locationt with_assertion_description(
```
Possibly useful, but please make this more generally available. On quick inspection almost every place that `git grep -n set_comment` highlights could benefit from this.

peterschrammel replied:
I don't think that makes the code more readable.

tautschnig commented on src/util/assert_or_die.h:
```diff
@@ -0,0 +1,16 @@
+
+#include "std_code.h"
+
+codet create_assert_or_die(const exprt &condition, const source_locationt &loc);
```
1) I think `expr_util` or `std_code` are perfectly good places to put this, no need for a new file.
2) Be specific about the return type. There is no need to hide the fact that it returns a `code_blockt`.
3) I'm not sure I like the name. How about `aborting_assert`?

smowton commented:
@tautschnig @peterschrammel changes applied, please re-review.

tautschnig commented on src/util/std_code.cpp:
```diff
@@ -107,3 +107,15 @@ void code_blockt::append(const code_blockt &extra_block)
+  result.copy_to_operands(code_assertt(condition));
+  result.copy_to_operands(code_assumet(condition));
+  for(auto &op : result.operands())
+    op.add_source_location() = loc;
```
How about:
```
code_blockt result;
result.add(code_assertt(condition), loc);
result.add(code_assumet(condition), loc);
```
or (leaving the loop intact)
```
code_blockt result({code_assertt(condition), code_assumet(condition)});
```

martin-cs reviewed (approved):
Seems in line with the general state of affairs.

smowton commented:
This has exposed some probably unrelated test failures. To investigate.

smowton committed:
Amend faulty long-string test

This test previously checked that the *sum* of the arguments' sizes didn't
exceed some critical bound, but this didn't exclude the case where one of them
is just within bounds (but still long enough to cause a memout when we try to
concretise it) and the other just long enough to push the sum over the boundary.

Instead we'll check that one or the other is very long, and the other one acceptably
short.

smowton commented:
The problem was another fragile string solver test that required some luck to test; it was failing because the md5sum of the binary changed / moon wasn't aligned with Saturn, rather than due to the actual change introduced here.. I've hardened the test, hopefully now this will pass CI.

smowton commented:
Bump: https://github.com/diffblue/test-gen/pull/1830

kroening commented:
BTW, related to this, I would be strongly in favour of making --java-throw-runtime-exceptions the default for jbmc. The Java semantics is very clear on these, this isn't the same situation as in C.

There can be a --java-runtime-exceptions-block or the like.

smowton commented:
@kroening as a prereq to --java-throw-runtime-exceptions we need to get a lot better at modelling control flow, otherwise a lot of implausible exceptional flows blur symex's certainty about the state of the world and the formula grows remarkably.

peterschrammel commented:
@smowton, could using https://github.com/diffblue/cbmc/blob/develop/src/analyses/goto_check.cpp#L1478 improve the situation (at least for NPEs)?
